Psalm 6

greg0ire opened the pull request:
<!-- Fill in the relevant information below to help triage your pull request. -->

|      Q       |   A
|------------- | -----------
| Type         | improvement
| BC Break     | no
| Fixed issues | n/a

#### Summary

Pushing although there are still 9 errors left, because they are either blocked, or I need help.
See:
- [x] https://github.com/vimeo/psalm/pull/3221
- [x] https://github.com/vimeo/psalm/issues/3224
- [x] https://github.com/vimeo/psalm/issues/3217
- [x] https://github.com/sebastianbergmann/phpunit/pull/4187
- [x] https://github.com/sebastianbergmann/phpunit/pull/4184
- [x] Rebase on top of https://github.com/doctrine/dbal/pull/3983 , use array shapes directly in `@param`
- [x] Figure out what to do about https://github.com/doctrine/dbal/pull/3977#discussion_r415032908
- [x] Figure out what to do about https://github.com/doctrine/dbal/pull/3977#issuecomment-619357227
- [x] Check if things can be improved with PHPStan following https://github.com/phpstan/phpstan/issues/3228 => can't reproduce now, not sure why
- [x] Wait for a stable release of Psalm
- [x] Wait for a stable release of PHPUnit
- [x] squash what makes sense squashing
- [x] do the merge locally
- [x] prepare [merge up for 2.11.x](https://github.com/doctrine/dbal/pull/4022)
- [x] prepare [merge up for 3.0.x](https://github.com/doctrine/dbal/pull/4025)
- [x] prepare [merge up for master](https://github.com/doctrine/dbal/pull/4026)

I will of course squash many of the commits together in the end.

greg0ire committed:
Document actual types

greg0ire committed:
Install phpunit plugin for Psalm

greg0ire committed:
Make return type verified and accurate

greg0ire committed:
Improve phpdoc

greg0ire committed:
Update phpunit

The latest version is easier to understand for static analyzers.

greg0ire committed:
Remove uneeded key

greg0ire committed:
Use list

It is more simple and more accurate

greg0ire committed:
Use shortcut

greg0ire committed:
Suppress deliberate error

We are testing what happens when providing the wrong type.

greg0ire committed:
Use TrimMode::UNSPECIFIED over null

The method signature does not allow null.

greg0ire committed:
Pass the right type of argument

greg0ire committed:
Ignore issues when testing with deliberate

greg0ire committed:
Make sure to pass in a string

Not all RDBMS have a concept of database, apparently (see SQLAnywhere).
Those who do not should pass in a string to stay compatible with the
signature.

greg0ire committed:
Document platform

greg0ire committed:
Add missing return type declarations

greg0ire committed:
Document null as a possible type

greg0ire committed:
Update PHPStorm stubs

That alone fixes 6 issues found by Psalm

greg0ire committed:
Use the appropriate notation for generators

greg0ire committed:
Document property

greg0ire committed:
Address deprecation

greg0ire committed:
Account for columnar expected result

This is what is expected with FETCH_COLUMN

greg0ire committed:
Fix wrong phpdoc

greg0ire committed:
Suppress InvalidArgument error

This is precisely what we are testing here.

greg0ire commented on composer.json:
```diff
@@ -42,7 +42,8 @@
         "nikic/php-parser": "^4.4",
         "phpstan/phpstan": "^0.12",
-        "phpunit/phpunit": "^8.4.1",
+        "phpunit/phpunit": "^8.5@dev",
```
Let's wait for https://github.com/sebastianbergmann/phpunit/pull/4187

greg0ire commented on lib/Doctrine/DBAL/Schema/ForeignKeyConstraint.php:
```diff
@@ -56,11 +56,11 @@ class ForeignKeyConstraint extends AbstractAsset implements Constraint
-     * @param mixed[]      $options            Options associated with the foreign key constraint.
+     * @param string[]           $localColumnNames   Names of the referencing table columns.
+     * @param Table|string       $foreignTableName   Referenced table.
+     * @param array<string|null> $foreignColumnNames Names of the referenced table columns.
```
This is the one I'm really not sure about. I wrote it because of this test:
https://github.com/doctrine/dbal/blob/155d028be084ef69110d09938c57c351a2126e5d/tests/Doctrine/Tests/DBAL/Functional/Schema/SqliteSchemaManagerTest.php#L86-L104

But it causes other issues down the line… I think there might be a bug here.

morozov replied:
In https://github.com/doctrine/dbal/pull/2854, `null` was replaced with an empty string.

greg0ire commented:
In the remaining issues, there's one I don't know what to do about:

```
ERROR: NullArgument - tests/Doctrine/Tests/DBAL/Functional/Driver/SQLSrv/StatementTest.php:30:68 - Argument 1 of Doctrine\DBAL\Driver\Connection::prepare cannot be null, null value provided to parameter with type string (see https://psalm.dev/057)
        $stmt = $this->connection->getWrappedConnection()->prepare(null);
```

https://github.com/doctrine/dbal/blob/155d028be084ef69110d09938c57c351a2126e5d/tests/Doctrine/Tests/DBAL/Functional/Driver/SQLSrv/StatementTest.php#L10-L37

Please help.

greg0ire committed:
Use an empty string instead of null

See https://github.com/vimeo/psalm/issues/3224

morozov commented on lib/Doctrine/DBAL/Connection.php:
```diff
@@ -1034,9 +1033,9 @@ public function query()
-     * @param int[]|string[] $types  The parameter types.
+     * @param string                 $query  The SQL query.
+     * @param mixed[]                $params The query parameters.
+     * @param array<int|string|null> $types  The parameter types.
```
Should `$params` and `$types` have the same "container" type since they both describe parameters? I.e. `mixed[]` → `array<mixed>`.

morozov commented on lib/Doctrine/DBAL/Driver/DrizzlePDOMySql/Driver.php:
```diff
@@ -34,15 +34,15 @@ public function createDatabasePlatformForVersion($version)
      * {@inheritdoc}
      */
-    public function getDatabasePlatform()
+    public function getDatabasePlatform() : DrizzlePlatform
```
Isn't it a breaking change?

greg0ire replied:
Woops it is indeed, since the class isn't final… I will use `@return` phpdoc instead.

morozov commented on lib/Doctrine/DBAL/Platforms/AbstractPlatform.php:
```diff
@@ -787,7 +787,7 @@ public function getModExpression($expression1, $expression2)
      *
      * @param string      $str  The expression to apply the trim to.
-     * @param int         $mode The position of the trim (leading/trailing/both).
+     * @param int|null    $mode The position of the trim (leading/trailing/both).
```
Why null? It doesn't seem to have any special meaning here.

greg0ire replied:
I thought having to _specify_ `TrimMode::UNSPECIFIED` would be too ironic. The purpose is to specify a 3rd argument.

```
ERROR: NullArgument - tests/Doctrine/Tests/DBAL/Platforms/SQLAnywherePlatformTest.php:611:58 - Argument 2 of Doctrine\DBAL\Platforms\SQLAnywherePlatform::getTrimExpression cannot be null, null value provided to parameter with type int (see https://psalm.dev/057)
            $this->platform->getTrimExpression('column', null, 'c')
```

morozov replied:
The test is obviously wrong. The method doesn't accept NULL. There's no way in PHP to omit the 2nd optional argument and specify the 3rd one.

greg0ire replied:
Ok I'll use `TrimMode::UNSPECIFIED`, I just noticed it is used in the rest of the method.

morozov commented on lib/Doctrine/DBAL/Platforms/AbstractPlatform.php:
```diff
@@ -2778,7 +2778,7 @@ public function getListNamespacesSQL()
 
     /**
-     * @param string $database
+     * @param string|null $database
```
Why null? None of the implementations expect null.

greg0ire replied:
```
ERROR: NullArgument - tests/Doctrine/Tests/DBAL/Platforms/SQLAnywhere12PlatformTest.php:55:50 - Argument 1 of Doctrine\DBAL\Platforms\SQLAnywhere12Platform::getListSequencesSQL cannot be null, null value provided to parameter with type string (see https://psalm.dev/057)
            $this->platform->getListSequencesSQL(null)
```


Looks like SQLAnywhere12 (at least) does https://github.com/doctrine/dbal/blob/155d028be084ef69110d09938c57c351a2126e5d/lib/Doctrine/DBAL/Platforms/SQLAnywhere12Platform.php#L65-L68

morozov replied:
Those methods ignore the passed value. It doesn't mean that they accept `NULL`.

greg0ire replied:
Ok, I will use `''` in tests that use `null` here.

greg0ire replied:
If that feels wrong, maybe we should use `?string $database` when merging up.

morozov replied:
`''` is better, it's at least of the correct type. `?string $database` is not an option. None of the implementations accept an optional DB name in `master` and this is done deliberately.

morozov commented on lib/Doctrine/DBAL/Platforms/AbstractPlatform.php:
```diff
@@ -2837,7 +2837,7 @@ public function getListUsersSQL()
      * Returns the SQL to list all views of a database or user.
      *
-     * @param string $database
+     * @param string|null $database
```
Why null?

greg0ire replied:
For the same reason, see an example among many: https://github.com/doctrine/dbal/blob/6c789b30a46906c5e669158809076d2405ea4fa9/lib/Doctrine/DBAL/Platforms/DB2Platform.php#L337-L340

morozov commented on tests/Doctrine/Tests/DBAL/Driver/StatementIteratorTest.php:
```diff
@@ -45,6 +46,8 @@ public function testIteratorIterationCallsFetchOncePerStep() : void
     }
 
     /**
+     * @param class-string<IteratorAggregate &Statement> $class
```
The whitespace before `&` looks confusing. `ResultStatement` already extends `Traversable`. Is `IteratorAggregate` really required here?

greg0ire replied:
I didn't mean to add the whitespace, sorry. I'll look into it

morozov commented on tests/Doctrine/Tests/DBAL/Driver/StatementIteratorTest.php:
```diff
@@ -79,9 +82,9 @@ private function assertIterationCallsFetchOncePerStep(Traversable $iterator, int
+     * @return Generator<class-string<IteratorAggregate&Statement>[]>
      */
-    public static function statementProvider() : iterable
+    public static function statementProvider() : Generator
```
Why does it have to be this specific? `Generator` is an implementation of `iterable` of which the caller doesn't care.

`class-string...[]` is misleading. We're not returning a list of class strings. We're returning a tuple, the first argument of which is a class-string. Adding another argument of another type to the tuple will break this expectation.

greg0ire replied:
Ok, I'll try another syntax which is less misleading. As for being specific, I usually use as specific return types as I can, it can't hurt, but I will change it back.

morozov replied:
> I usually use as specific return types as I can

I don't think it's the end goal. The test methods implement the PHPUnit's API that expects data providers to [return a `Traversable`](https://github.com/sebastianbergmann/phpunit/blob/85fefa381541fb3fcbbfe9b5f32be386551cd774/src/Util/Annotation/DocBlock.php#L412). It is fine to be more specific about the contained data type because you want to make sure the items are correct. But there's no point in being more specific about the container type unless you want to create extra coupling between the method signature (`fn (): array`/`fn (): Generator`) and the implementation (`return [];`/`yield ...` respectively).

morozov commented on tests/Doctrine/Tests/DBAL/Platforms/AbstractMySQLPlatformTestCase.php:
```diff
@@ -207,6 +207,8 @@ public function testGetDateTimeTypeDeclarationSql() : void
     /**
      * {@inheritDoc}
+     *
+     * @psalm-return array
```
Not sure what this means.

greg0ire replied:
The parent has `@psalm-return never-return`, hence this comment.

morozov commented on tests/Doctrine/Tests/DBAL/Platforms/AbstractPlatformTestCase.php:
```diff
@@ -535,6 +535,8 @@ public function testCreateTableColumnTypeComments() : void
     /**
      * @return string[]
+     *
+     * @psalm-return never-return
```
Should this be fixed in PHPUnit or in Psalm? If a function unconditionally calls a method that is marked as never-return (first option) or always throws an exception (second option), it should be automatically detected as such.

greg0ire replied:
I think I already fixed everything there is to fix in PHPUnit: https://github.com/sebastianbergmann/phpunit/pull/4187

As for Psalm, I don't know… @muglug, sorry to invoke you here, but can you please comment on this? :pray:

morozov commented on tests/Doctrine/Tests/DBAL/Types/TypeTest.php:
```diff
@@ -28,7 +30,10 @@ public function defaultTypesProvider() : iterable
 
-            yield [$constant->getValue()];
+            $constantValue = $constant->getValue();
+            assert(is_string($constantValue));
```
Should we use PHPUnit assertions in tests? They are more specific and are always effective regardless of the runtime configuration.

greg0ire replied:
Sure

greg0ire committed:
Use level 6

greg0ire commented on lib/Doctrine/DBAL/DriverManager.php:
```diff
@@ -115,11 +115,15 @@ private function __construct()
+     * @param Configuration|null                    $config       The configuration to use.
+     * @param EventManager|null                     $eventManager The event manager to use.
+     *
+     * @return T
```
I'm going to revert this, because if you pass in `null` as a `wrapperClass`, then it returns a `Connection`, and I don't think there is way to express that it return `Connection` if null, or `T` otherwise.

morozov replied:
Would it help if we didn't ignore `NULL and used `array_key_exists()` instead of `isset()`. API-wise, what's the point in passing an option with `NULL` and expecting that it will be ignored?

greg0ire replied:
I just tried, and it does not help Psalm. These errors come back as soon as I add the `|Connection`, and using `array_key_exists` does not improve that situation:

```
ERROR: UndefinedMethod - tests/Doctrine/Tests/DBAL/Sharding/PoolingShardConnectionTest.php:193:33 - Method Doctrine\DBAL\Connection::getactiveshardid does not exist (see https://psalm.dev/022)
        self::assertNull($conn->getActiveShardId());


ERROR: UndefinedMethod - tests/Doctrine/Tests/DBAL/Sharding/PoolingShardConnectionTest.php:196:38 - Method Doctrine\DBAL\Connection::getactiveshardid does not exist (see https://psalm.dev/022)
        self::assertEquals(0, $conn->getActiveShardId());


ERROR: UndefinedMethod - tests/Doctrine/Tests/DBAL/Sharding/PoolingShardConnectionTest.php:199:38 - Method Doctrine\DBAL\Connection::getactiveshardid does not exist (see https://psalm.dev/022)
        self::assertEquals(1, $conn->getActiveShardId());


ERROR: UndefinedMethod - tests/Doctrine/Tests/DBAL/Sharding/PoolingShardConnectionTest.php:202:33 - Method Doctrine\DBAL\Connection::getactiveshardid does not exist (see https://psalm.dev/022)
        self::assertNull($conn->getActiveShardId());
```

greg0ire replied:
I just asked for support here, just in case: https://symfony-devs.slack.com/archives/C8SFXTD2M/p1588144518359100

greg0ire replied:
This is going to get fixed: https://github.com/vimeo/psalm/issues/3277 :)

greg0ire replied:
I have it working locally, but now PHPStan is unhappy about it :see_no_evil:

greg0ire replied:
I pushed a new commit that include additional rules to ignor ethe PHPStan issues this creates.

greg0ire commented:
@morozov the failure in Travis is caused by an issue that is fixed on 2.11 because `VersionAwarePlatformDriver` extends `Driver` on that branch. Should I create a short-lived test interface that extends both and use it in the test? I would remove it when merging up.

morozov commented:
> Should I create a short-lived test interface that extends both and use it in the test?

That should be fine.

greg0ire committed:
Describe return type accurately

This return type is conditional to $params['wrapperClass']. Luckily,
recent versions of Psalm allow documenting it properly.
See https://github.com/vimeo/psalm/issues/3277

Note that phpstan is not able to understand this yet, but still attempts
to, hence the extra ignore rules.

greg0ire committed:
Address warning from PHPUnit

This hack is temporary and should be removed on 3.0.x

greg0ire commented:
@morozov phpcs fails unless I use `@psalm-param` instead of `@param` when using array shapes. An alternative is to get `slevomat/coding-standard` to 6.3.3, which can be done by upgrading `phpstan/phpdoc-parser`, `doctrine/coding-standard` and `squizlabs/php_codesniffer`. Should I go ahead and create another PR that upgrades these packages? I might make other things easier.

morozov commented on lib/Doctrine/DBAL/DriverManager.php:
```diff
@@ -115,11 +115,16 @@ private function __construct()
      * @throws DBALException
+     *
+     * @phpstan-param mixed[] $params
+     * @phpstan-return Connection
```
This is getting ugly. You need to tell each SA what's happening in its own language. I hope it will force us to introduce simpler and cleaner APIs.

morozov replied:
Why is `@phpstan-return` needed? There's no `@return` and there is a return type declaration with the same type. Does PHPStan get confused with `@psalm-return` and needs an override?

greg0ire replied:
Removing the first one causes this:

```
 ------ ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 
  Line   Doctrine/DBAL/Id/TableGenerator.php                                                                                                                                               
 ------ ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 
  75     Parameter #1 $params of static method Doctrine\DBAL\DriverManager::getConnection() expects array()|array('wrapperClass' => class-string<Doctrine\DBAL\Connection>), array given.  
 ------ ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 

 ------ -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 
  Line   Doctrine/DBAL/Schema/SqliteSchemaManager.php                                                                                                                                                            
 ------ -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 
  59     Parameter #1 $params of static method Doctrine\DBAL\DriverManager::getConnection() expects array()|array('wrapperClass' => class-string<Doctrine\DBAL\Connection>), array('driver' => mixed, 'path' =>  
         string) given.                                                                                                                                                                                          
 ------ -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 
```
I will remove `@phpstan-return`, it is not needed, it's a remnant of when I was trying to get phpstan to be quiet about that warning I added in the ignore list.

morozov commented on lib/Doctrine/DBAL/DriverManager.php:
```diff
@@ -115,11 +115,16 @@ private function __construct()
-     * @param mixed[]            $params       The parameters.
-     * @param Configuration|null $config       The configuration to use.
-     * @param EventManager|null  $eventManager The event manager to use.
+     * @param array{wrapperClass?: class-string<T>} $params
```
Can we turn this the other way around and use `@psalm-param` for this definition in order to not distract PHPStan that doesn't understand this?

greg0ire replied:
I will check, but PHPStan and Psalm read each others' annotation

greg0ire replied:
As I feared, it does not work, I get the error messages I described at https://github.com/doctrine/dbal/pull/3977#discussion_r418601811

morozov replied:
> I will check, but PHPStan and Psalm read each others' annotation

If they read each other's annotations but don't interpret equally then what's the point in that and having the `psalm-` and `phpstan-` prefixes? The difference in interpretation shouldn't be solved here.

greg0ire replied:
I think the point is not to confuse other tools like phpcs+slevomat, and that they started with not interpreting each other's rules at the beginning.

morozov commented:
@greg0ire let's make the needed upgrades and see if we can simplify the annotations. The ones in the driver manager are really concerning.

greg0ire committed:
Add contract for createPlatform()'s return type

greg0ire commented:
Down to 4 errors.

greg0ire commented:
Down to 2 errors, with which I need help please @morozov :pray: see the todo list in the original post for details.

morozov commented:
> https://github.com/doctrine/dbal/blob/155d028be084ef69110d09938c57c351a2126e5d/tests/Doctrine/Tests/DBAL/Functional/Driver/SQLSrv/StatementTest.php#L29-L30

IIRC, it just needs to be an invalid SQL that will cause failure being prepared. An empty string instead of `null` should work.

greg0ire committed:
Add missing use statement

greg0ire committed:
Use valid arguments

morozov commented on lib/Doctrine/DBAL/Driver/PDOSqlsrv/Connection.php:
```diff
@@ -33,8 +35,10 @@ public function lastInsertId($name = null)
         $stmt = $this->prepare('SELECT CONVERT(VARCHAR(MAX), current_value) FROM sys.sequences WHERE name = ?');
         $stmt->execute([$name]);
+        $lastInsertId = $stmt->fetchColumn();
+        assert(is_string($lastInsertId));
```
This doesn't look valid. If the sequence doesn't exist, the query will return `false`.

greg0ire replied:
Not sure why I did this, reverting it doesn't make Psalm fail :thinking:

greg0ire replied:
Ah wait that was Phpstan:

```
 ------ ------------------------------------------------------------------ 
  Line   Doctrine/DBAL/Driver/PDOSqlsrv/Connection.php                     
 ------ ------------------------------------------------------------------ 
  37     Method Doctrine\DBAL\Driver\PDOSqlsrv\Connection::lastInsertId()  
         should return string but returns string|false|null.               
 ------ ------------------------------------------------------------------ 
```

greg0ire replied:
Ok I will assert that it's either a string or `false`

morozov replied:
It's better to assert that it's [not null](https://docs.microsoft.com/en-us/sql/relational-databases/system-catalog-views/sys-sequences-transact-sql?view=sql-server-ver15) (see the column type).

greg0ire replied:
This might be an LSP violation actually: this method should only ever return strings according to [the docs](https://www.php.net/manual/en/pdo.lastinsertid.php) (and also the phpstorm stubs)
Should we ignore this particular violation? Or should we return `'0'` if not a string?

morozov replied:
The documentation on PHP is really vague because it doesn't describe any non-happy paths except for "not implemented". Since it's `2.x`, let's suppress the error instead of modifying the behavior in any way.

SenseException commented on tests/Doctrine/Tests/DbalPerformanceTestListener.php:
```diff
@@ -16,7 +16,7 @@ class DbalPerformanceTestListener implements TestListener
     use TestListenerDefaultImplementation;
 
-    /** @var string[][] */
+    /** @var float[][] */
```
```suggestion
    /** @var array<string, array<string, float>> */
```

greg0ire replied:
Fixed

SenseException commented on tests/Doctrine/Tests/DBAL/Platforms/OraclePlatformTest.php:
```diff
@@ -356,6 +362,8 @@ public function testGenerateTableWithAutoincrement() : void
     /**
      * {@inheritDoc}
+     *
+     * @psalm-return array
```
Does Psalm complain about the return type given in the parent?

greg0ire replied:
It does complain in my IDE, maybe that's a complain at a higher level? Anyway, I think I am going to remove this and use `never-return|string[]` in the parent, see https://github.com/phpstan/phpstan/issues/3335 . WDYT?

greg0ire replied:
In fact, the only version I could get to work on both most recent versions of both tools is that one:
- https://psalm.dev/r/01f47ec513
- https://phpstan.org/r/87a19063-3ef0-4a8a-9c1c-301d662d0090

greg0ire replied:
Maybe I should report this: https://psalm.dev/r/01f47ec513

greg0ire replied:
I asked for help at https://symfony-devs.slack.com/archives/C8SFXTD2M/p1590518449028600

greg0ire replied:
Ok, here is a commit that should help with Psalm https://github.com/vimeo/psalm/commit/f0a5bd74b6f81c40393d45cd3490a5d2721509c2 :)

greg0ire replied:
I updated Psalm, and I can drop the commit about never return :tada: 

But now there are new issues to fix :see_no_evil:

SenseException replied:
It was expected after an update :smile: 

Nice work.

morozov commented:
@greg0ire before this PR gets merged, does it need to be 33 commits long? I'd rather see the commits grouped into logical blocks that provide a bigger picture of why certain changes are made rather than “did this”, “did that”.

greg0ire commented:
> @greg0ire before this PR gets merged, does it need to be 33 commits long? I'd rather see the commits grouped into logical blocks that provide a bigger picture of why certain changes are made rather than “do this”, “do that”.

The big picture is : "fix issues detected by Psalm". That consists in many, many unrelated things, because I'm fixing 136 errors here. Some of them are trivial, some others require a github link, or an explanation, and should not be grouped with others.

morozov commented:
@greg0ire since there's some work to do, can we do up-merges separately? Specifically, I'm interested in getting `2.11.x` → `3.0.x` to proceed with the removal of the deprecated fetch-related APIs.

greg0ire commented:
> @greg0ire since there's still some work to do, can we do up-merges separately? Specifically, I'm interested in getting 2.11.x → 3.0.x to proceed with the removal of the deprecated fetch-related APIs.

Yes, sure, my plan of using `--atomic` was a bit unrelatistic I'm afraid. I'm leaving my PRs open because some contains commits that are specific to some branches (new code requiring additional fixes). I don't regret creating them preemptively though, because now I have a better idea of the amount of work required to merge this up.

greg0ire committed:
Update psalm

greg0ire committed:
Avoid mocking inexistent classes

It confuses both Psalm and PHPStan, and involves complicated mocking.

greg0ire commented:
@morozov @SenseException I pushed a new version that drops the "never-return" commit, it is no longer needed with the latest version of Psalm. I had to push a commit that avoids mocking interfaces that do not exist, I'm not sure you're going to like it but it's either that or ignoring these errors, tell me what you prefer.

SenseException reviewed (dismissed):
I'm in favour of the interfaces you provided. These are visible and can be removed once there is another solution for this. Does it makes sense to move those in separate files?

greg0ire commented:
> Does it makes sense to move those in separate files?

They don't need to be autoloaded from other test files, so I would say no, since it means less clutter in the directory.

greg0ire committed:
Ignore error about Connection::lastInsertId()

The php docs are not very clear about that one.

morozov commented on tests/Doctrine/Tests/DBAL/FutureVersionAwarePlatformDriver.php:
```diff
@@ -0,0 +1,15 @@
+use Doctrine\DBAL\VersionAwarePlatformDriver;
+
+/**
+ * Remove me in 2.11
```
Should it be “in 3.0.x"? See https://github.com/doctrine/dbal/pull/3903.

morozov commented on tests/Doctrine/Tests/DBAL/Platforms/AbstractPlatformTestCase.php:
```diff
@@ -1532,3 +1518,28 @@ public function testZeroOffsetWithoutLimitIsIgnored() : void
     }
 }
+
+interface GetCreateTableSqlDispatchEventListener
```
I hope one day it gets fixed in the event dispatcher so that instead of just calling random methods dynamically, it lets one define an interface and accepts its implementors as listeners. Otherwise, it looks barely manageable.

morozov commented on tests/Doctrine/Tests/DBAL/Platforms/SQLAnywherePlatformTest.php:
```diff
@@ -875,6 +881,8 @@ protected function getQuotedAlterTableRenameColumnSQL() : array
     /**
      * {@inheritdoc}
+     *
+     * @psalm-return never-return
```
Is it sill needed?

greg0ire replied:
It isn't, I'll remove it, good catch!

greg0ire commented:
![Finally!](https://i.redd.it/0ah2bv3ihy7y.gif)

pkruithof commented:
Sorry to comment on an older PR, but this commit causes phpstan to report an error: https://phpstan.org/r/e462e0f9-abc4-44bc-8a9a-80aca2717980

I've asked a support question there, and the author suggested a solution here: https://github.com/phpstan/phpstan/issues/3849#issuecomment-689517922

Would you accept a PR with the second solution? I don't think a third-party library should trigger phpstan errors in a project.

morozov commented:
@pkruithof please go ahead and submit a PR. @greg0ire could you work with PHPStan on finding the root cause as suggested in the first option?

pkruithof commented:
I just tried the suggestion but it did not help unfortunately, and I'm kind of out of my depth here. 😞

morozov commented:
@pkruithof let's start with filing a new issue that would reproduce the problem.

pkruithof commented:
Done 👉 #4264
